### PR TITLE
add mdadm support

### PR DESCRIPTION
**Pull request by SourceDoctor**
Feature to monitor mdadm
watching this attributes:
- Raid Level
- Raid Size
- Raid Device Count
- Hotspare Count
- degraded state
- sync Speed
- sync completed

### CODE REVIEW

**Comment by SourceDoctor** [comment]
Application Parts for Mdadm:

**SNMP Extends:**
mdadm.conf:
`# reading mdadm
extend mdadm /etc/snmp/mdadm
`
**mdadm snmp Configfile**
mdadm.config:
`SUDO=/usr/bin/sudo
SMARTCTL=/usr/sbin/smartctl
AWK=/usr/bin/awk
CAT=/bin/cat
LS=/bin/ls
ECHO=/bin/echo
`

**mdadm snmp Script**
mdadm:
`#!/bin/bash

CONFIGFILE=$0.config

 . $CONFIGFILE

if [ -d /dev/md ] ; then
    for RAID in /sys/block/md* ; do

        # ignore arrays with no slaves
        if [ -z "$($LS -1 $RAID/slaves)" ] ; then
            continue
        fi
        # ignore "non existing" arrays
        if [ ! -f "$RAID/md/degraded" ] ; then
            continue
        fi

        RAID_NAME=$(basename $RAID)
        _RAID_DEV_LIST=$($LS $RAID/slaves/)
        RAID_LEVEL=$($CAT $RAID/md/level)
        RAID_DISC_COUNT=$($CAT $RAID/md/raid_disks)
        RAID_STATE=$($CAT $RAID/md/array_state)
        RAID_ACTION=$($CAT $RAID/md/sync_action)
        RAID_DEGRADED=$($CAT $RAID/md/degraded)
        RAID_SYNC_SPEED=$($CAT $RAID/md/sync_speed)
        if [ "$RAID_SYNC_SPEED" = "none" ] ; then
            RAID_SYNC_SPEED=0
        fi
        if [ "$($CAT $RAID/md/sync_completed)" = "none" ] ; then
            RAID_SYNC_COMPLETED=100
        else
            let "RAID_SYNC_COMPLETED=100*$($CAT $RAID/md/sync_completed)"
        fi
        # divide with 2 to size like in /proc/mdstat
        let "RAID_SIZE=$($CAT $RAID/size)/2"

        RAID_MISSING_DEVICES=""
        for D in $_RAID_DEV_LIST ; do
            if [ -L $RAID/slaves/$D ] && [ -f $RAID/slaves/$D ] ; then
                RAID_MISSING_DEVICES=$RAID_MISSING_DEVICES" "$D
            fi
        done

        RAID_DEV_LIST=$_RAID_DEV_LIST
        RAID_MISSING_DEVICES=$_RAID_MISSING_DEVICES

        $ECHO -n $RAID_NAME","
        $ECHO -n $RAID_LEVEL","
        $ECHO -n $RAID_SIZE","
        $ECHO -n $RAID_DISC_COUNT","
        $ECHO -n $RAID_DEV_LIST","
        $ECHO -n $RAID_MISSING_DEVICES","
        $ECHO -n $RAID_STATE","
        $ECHO -n $RAID_ACTION","
        $ECHO -n $RAID_DEGRADED","
        $ECHO -n $RAID_SYNC_SPEED","
        $ECHO -n $RAID_SYNC_COMPLETED
        $ECHO ""
    done
fi
`

**Comment by murrant** [comment]
You should send a PR for the extend script to https://github.com/librenms/librenms-agent
Also, the extend should output json, see some of the other scripts for examples.

**Comment by SourceDoctor** [comment]
mdadm Monitoring now expect's JSON
snmp Client can be found here:
https://github.com/librenms/librenms-agent/pull/230

**Comment by SourceDoctor** [comment]
@murrant code is fixed :-)

File: app/Models/Application.php [hunk 1]
```diff
@@ -49,6 +49,7 @@ class Application extends DeviceRelatedModel
         'fbsd-nfs-server' => 'FreeBSD NFS Server',
         'freeradius' => 'FreeRADIUS',
         'gpsd' => 'GPSD',
+        'mdadm' => 'Mdadm',
```
Comment (murrant) [inline app/Models/Application.php]: This is the default capitalization, we can remove this line.

File: includes/html/pages/device/apps/mdadm.inc.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    'mdadm_sync_speed'     => 'RAID Sync speed',
+    'mdadm_sync_completed' => 'RAID Sync completed',
+);
+
+
+foreach ($graphs as $key => $text) {
+    $graph_type            = $key;
+    $graph_array['height'] = '100';
+    $graph_array['width']  = '215';
+    $graph_array['to']     = $config['time']['now'];
```
Comment (murrant) [inline includes/html/pages/device/apps/mdadm.inc.php]: Remove use of $config.  Use Config::get('time.now') or time()

**Comment by murrant** [comment]
Looks mostly ready, just a few minor cleanups.  Also can you change array() -> []?

**Comment by SourceDoctor** [comment]
@murrant 
i changed the code like you wished in review. it can be merged now hopefully ;-)

**Comment by murrant** [comment]
Still have a lot of array() calls, but I'm not overly concerned about that just trying to reduce future code churn when we change all definitions over.

**Comment by SourceDoctor** [comment]
i've got array calls, because several RAID Arrays are possible ;-)
So usage of an array for it is a common way.

**Comment by SourceDoctor** [comment]
@murrant something else to rewrite?

**Comment by corsoblaster** [comment]
No other about this topic.

I would like also to replace all related to Cyberoam with Sophos, because Cyberoam does’nt exists anymore. But it is complex, due to the fact that there is a different MIB file, I am new in Github / SNMP / MIB etc etc

 

I’ll do one thing at time, starting to replace logo and icon. I thought to open a different pull request. Is it correct?

 

 

 

Da: SourceDoctor <notifications@github.com>
Risposta: librenms/librenms <reply@reply.github.com>
Data: domenica 12 maggio 2019 16:04
A: librenms/librenms <librenms@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Oggetto: Re: [librenms/librenms] add mdadm support (#10159)

 

@murrant something else to rewrite?

**Pull request merged**